src/shared/read-json-file.ts

```diff
@@ -0,0 +1,31 @@
+import { promiseWithResolvers } from './promise-with-resolvers';
+
+/**
+ * Reads and tries to JSON-parse a file
+ * @return - Promise wrapping parsed content or rejected with any error that
+ *           occurred during parsing
+ */
+export function readJSONFile(
+  file: File,
+  /* istanbul ignore next - test seam */
+  fileReader = new FileReader(),
+): Promise<unknown> {
+  const { promise, resolve, reject } = promiseWithResolvers();
+
+  fileReader.addEventListener(
+    'loadend',
+    ({ target }: ProgressEvent<FileReader>) => {
+      try {
+        const content = target?.result?.toString() ?? '';
```

Comment:
OK, the event target here is the `FileReader` instance itself. And `result`'s type will depend on how the `FileReader` instance's read was done. In this case, you've used `readAsText`, so I'm curious about the `toString` here.

From MDN:

> The readAsText() method is used to read the contents of the specified [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or [File](https://developer.mozilla.org/en-US/docs/Web/API/File). When the read operation is complete, the [readyState](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readyState) is changed to DONE, the [loadend](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/loadend_event) event is triggered, and **the [result](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/result) property contains the contents of the file as a text string.**

Reply:
OK, second question, more high-level. In this logic, if either `target` or `result` are nullish, `content` will be set to the empty string. What does this state represent? How might this occur? I'm not suggesting you reject the Promise in this case necessarily (as rejections should only happen if the operation itself encountered an unexpected error), just curious.

Reply:
> OK, the event target here is the FileReader instance itself. And result's type will depend on how the FileReader instance's read was done. In this case, you've used readAsText, so I'm curious about the toString here.

At runtime we know this is going to be a string, but on types level, this is typed as `string | ArrayBuffer`. Using `toString()` is a simple way to narrow down the type to `string`.

![image](https://github.com/hypothesis/client/assets/2719332/38cb2421-6c52-4bc9-b988-8d4cec2ad692)

Reply:
> In this logic, if either target or result are nullish, content will be set to the empty string. What does this state represent? How might this occur?

This is again probably a problem with types, which do not clearly state when could this happen, but apparently it can.

Falling back to `''` makes `JSON.parse` fail, resulting in a rejected promise. However, this is probably not obvious enough.

I think we have three options here

1. We tell TS we know that's not going to be `undefined` by doing `e.target!.result!.toString()`, and let the `catch` capture if it happens to be undefined at runtime.
2. We do a previous check of content, making sure it is defined, and rejecting the promise with a known error. This is more verbose but easier to reason about when read.
    ```ts
    const content = e.target?.result?.toString();
    if (!content) {
      throw new Error('It was not possible to read file');
    }
  
    const parsed = JSON.parse(content);
    // [...]
    ```